packages/modules/vehicles/smarteq/socutils.py

```diff
@@ -0,0 +1,221 @@
+    global logFilter
+    if filter in logFilter or 'A' in logFilter:
+        log.info('(' + filter + '): ' + txt)
+
+
+def mqtt_on_connect(client, userdata, flags, rc):
+    global threadname
+    threading.current_thread().name = threadname
+    try:
+        if rc != 0:
+            log.error('mqtt_on_connect: failure: ' + str(rc))
+        client.subscribe("openWB/vehicle/+/soc_module/config")
+    except Exception as e:
+        log.exception("mqtt_on_connect error: %s", e)
+
+
+def mqtt_on_message(client, userdata, msg):
+    global threadname
+    global vehicleList
+    global vehicleConfig
+    try:
+        threading.current_thread().name = threadname
+        ev = msg.topic.split('/')[2]
+        m = json.loads(str(msg.payload, 'utf-8'))
+        if 'type' in m and m['type'] == 'smarteq':
+            vin = m['configuration']['vin']
+            _infoLog('Q', "mqtt_on_message smarteq: ev=" + ev + ', config =\n' + json.dumps(m, indent=4))
+            vehicleConfig[ev] = m
+            vehicleList[vin] = ev
+    except Exception as e:
+        log.exception("mqtt_on_message error: %s", e)
+
+
+class socUtils:
+
+    def __init__(self):
+        global threadname
+        global logFilter
+        logFilter = 'q'
+        threadname = 'soc_smarteq_mqtt'
+
+    def get_vin_ev_map(self) -> dict:
+        vh_list = {}
+        # try to get vin->ev.num mapping from ev_data object
+        for ev in data.data.ev_data.values():
+            # log.info('ev_data: ev=' + str(ev.num) + ', soc_module=' + str(ev.soc_module))
+            if isinstance(ev.soc_module, ConfigurableVehicle):
+                cv: ConfigurableVehicle = ev.soc_module
+                # log.info('ev_data: ev=' + str(ev.num) + ', cv.vehicle=' + str(cv.vehicle))
+                if isinstance(cv.vehicle_config, SmartEQ):
+                    se: SmartEQ = cv.vehicle_config
+                    # log.info('ev_data: ev=' + str(ev.num) + ', se.name/type=' + str(se.name) + '/' + str(se.type))
+                    if isinstance(se.configuration, SmartEQConfiguration):
+                        sc: SmartEQConfiguration = se.configuration
+                        _infoLog('i', 'ev_data: ev=' + str(ev.num) + ', sc.vin=' + str(sc.vin))
+                        vh_list[sc.vin] = ev.num
+        return vh_list
+
+    def read_token_file(self, path: str) -> str:
+        try:
+            with open(path, "r") as tf:           # try to open Token file
+                token = tf.read()              # read token
+        except Exception:
+            token = None                # if no old token found set Token_old to dummy value
+        return token
+
+    def write_token_file(self, path: str, token: str, config={}):
+        try:
+            _infoLog('s', "store Token in file " + path)
+            with open(path, "w") as tf:
+                tf.write(token)         # write Token file
+        except Exception as e:
+            log.exception('Token file write exception ' + str(e))
```

Comment:
Die tokens werden im Broker gespeichert, warum nochmal in eine Datei?

Reply:
Im broker wird aktuell nur der refresh.token gespeichert.
In der Ramdisk-Datei werden alle vom Server gelieferten Daten gespeichert.
Bei smarteq/mercedes macht das wenig Unterschied, das refresh- und access-token die gleiche Gültigkeit von 2 Stunden haben.
Bei VWId ist der Access-Token auch recht kurz gültig, der refresh-token (JWT) dagegen gilt für ca. 6 Monate, d.h. im broker wird nur sehr selten gespeichert.